esp-hal/src/gpio/interconnect.rs

```diff
@@ -25,9 +22,41 @@ use crate::{
     private::{self, Sealed},
 };
 
+/// A signal that can be connected to a peripheral input.
+///
+/// Peripheral drivers are encouraged to accept types that implement this and
+/// [`PeripheralOutput`] as arguments instead of pin types.
+#[doc(hidden)]
+pub trait PeripheralInput: Into<InputConnection> + 'static {}
+
+/// A signal that can be connected to a peripheral input and/or output.
+///
+/// Peripheral drivers are encouraged to accept types that implement this and
+/// [`PeripheralInput`] as arguments instead of pin types.
+#[doc(hidden)]
+pub trait PeripheralOutput: Into<OutputConnection> + 'static {}
+
+impl<P: InputPin> PeripheralInput for P {}
+impl<P: OutputPin> PeripheralOutput for P {}
+
+impl PeripheralInput for InputSignal {}
+impl PeripheralInput for OutputSignal {}
```

Comment:
Why have `OutputSignal` and `OutputConnection` implement `PeripheralInput`?

I noticed in the drivers that used to ask for bidirectional signals (i.e. `impl PeripheralInput + PeripheralOutput`), they've been changed to now just ask for `PeripheralOutput`, and this impl here allows that to work, but I don't see/understand the motive for that.

I suppose there's no such thing as an output-only pin, so on the hardware level it doesn't matter.

I still liked being able to see what the driver actually needed though. input, output or both.

Reply:
> I suppose there's no such thing as an output-only pin, so on the hardware level it doesn't matter.

Exactly this. `PeripheralInput + PeripheralOutput` is just the same thing written in more words.

Reply:
Would a naming change help here? e.g. `PeripheralInputOutput`?

Reply:
I guess, but we also don't call our GPIO traits `InputOutputPin`

Reply:
I think we had naming like that before 🤔 but not anymore 😄